src/fields/field_model.hh

```diff
@@ -0,0 +1,196 @@
+#include "fields/field_common.hh"
+#include "fields/field_algo_base.hh"
+#include "fields/field_values.hh"
+#include "fields/field_value_cache.hh"
+
+template <int spacedim> class ElementAccessor;
+
+
+/**
+ * support for expanding tuples into an overloaded function call's arguments
+ */
+#define wrap_overload(func) [](auto&&... ps){ return func( std::forward<decltype(ps)>(ps)... ); }
+
+namespace detail
+{
+//    //
+//    // base case for building up arguments for the function call
+//    //
+//    template< typename CALLABLE, typename TUPLE, int INDEX >
+//    struct tuple_into_callable_n
+//    {
+//        template< typename... Vs >
+//        static auto apply(CALLABLE f, TUPLE t, Vs&&... args) -> decltype(auto)
+//        {
+//            return tuple_into_callable_n<CALLABLE, TUPLE, INDEX - 1>::apply(
+//                f,
+//                std::forward<decltype(t)>(t),
+//                std::get<INDEX - 1>(std::forward<decltype(t)>(t)),
+//                std::forward<Vs>(args)...
+//            );
+//        }
+//    };
+//
+//    //
+//    // terminal case - do the actual function call
+//    //
+//    template< typename CALLABLE, typename TUPLE >
+//    struct tuple_into_callable_n< CALLABLE, TUPLE, 0 >
+//    {
+//        template< typename... Vs >
+//        static auto apply(CALLABLE f, TUPLE t, Vs&&... args) -> decltype(auto)
+//        {
+//            return f(std::forward<Vs>(args)...);
+//        };
+//    };
+
+
+    //
+    // base case for building up arguments for the function call
+    //
+    template< typename CALLABLE, typename TUPLE, int INDEX >
+    struct model_cache_item
+    {
+        template< typename... Vs >
+        static auto eval(int i_cache, CALLABLE f, TUPLE t, Vs&&... args) -> decltype(auto)
+        {
+            return model_cache_item<CALLABLE, TUPLE, INDEX - 1>::eval(
+                i_cache,
+                f,
+                std::forward<decltype(t)>(t),
+                std::get<INDEX - 1>(std::forward<decltype(t)>(t))[i_cache],
+                std::forward<Vs>(args)...
+            );
+        }
+    };
+
+    //
+    // terminal case - do the actual function call
+    //
+    template< typename CALLABLE, typename TUPLE >
+    struct model_cache_item< CALLABLE, TUPLE, 0 >
+    {
+        template< typename... Vs >
+        static auto eval(int i_cache, CALLABLE f, TUPLE t, Vs&&... args) -> decltype(auto)
+        {
+            return f(std::forward<Vs>(args)...);
+        };
+    };
+
+
+}
+
+//template< typename FUNC, typename TUPLE >
+//auto tuple_into_callable(FUNC f, TUPLE&& t) -> decltype(auto)
+//{
+//    return
+//        detail::tuple_into_callable_n<
+//            FUNC,
+//            decltype(t),
+//            std::tuple_size< std::remove_reference_t<TUPLE> >::value
+//        >::apply(f, std::forward<decltype(t)>(t) );
+//}
+
+
+template<int spacedim, class Value, class Fn, class ... InputFields>
+class FieldModel : public FieldAlgorithmBase<spacedim, Value>
```

Comment:
Add a comment with a usage example:
1. declaration of a Field
2. set its algo base on all regionons to FieldModel using Model::create.